File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        m, n = len(matrix), len(matrix[0])
        self.acc = [[0] * n for _ in range(m)]
        for row in range(m):
            row_sum = 0
            for col in range(n):
                upper_rect_sum = 0 if row == 0 else self.acc[row - 1][col]
                row_sum += matrix[row][col]
                self.acc[row][col] = row_sum + upper_rect_sum

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        upper_rect = 0 if row1 == 0 else self.acc[row1 - 1][col2]
        left_rect = 0 if col1 == 0 else self.acc[row2][col1 - 1]
        diag_rect = 0 if col1 == 0 or row1 == 0 else self.acc[row1 - 1][col1 - 1]
        return self.acc[row2][col2] - upper_rect - left_rect + diag_rect

# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.cpp
class NumMatrix {
    vector<vector<int>> vvi;
public:
    NumMatrix(vector<vector<int>>& matrix) {
        int row = matrix.size(), col = matrix[0].size();
        vvi.reserve(row + 1);
        vvi.push_back(vector<int>(col + 1, 0));
        for (int i = 1; i <= row; ++i) {
            vector<int> vi(1, 0);
            vi.reserve(col + 1);
            vvi.push_back(vi);
            for (int j = 1; j <= col; ++j) {
                int pi = i - 1, pj = j - 1;
                vvi[i].push_back(matrix[pi][pj] + vvi[pi][j] + vvi[i][pj] - vvi[pi][pj]);
            }
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        return vvi[++row2][++col2] - vvi[row2][col1] - vvi[row1][col2] + vvi[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */